packages/studio/src/store/slices/pages/createPageSlice.ts

```diff
@@ -160,6 +160,82 @@ export const createPageSlice: SliceCreator<PageSlice> = (set, get) => {
         (component) => component.uuid === activeComponentUUID
       );
     },
+    addSelectedComponentUUID: (selectedUUID: string) => {
+      set((store) => {
+        store.selectedComponentUUIDs.add(selectedUUID);
+      });
+    },
+    addSelectedComponentRect: (
+      selectedUUID: string,
+      rect: DOMRectProperties
+    ) => {
+      set((store) => {
+        store.selectedComponentRectsMap.set(selectedUUID, rect);
+      });
+    },
+    clearSelectedComponents: () => {
+      set((store) => {
+        store.selectedComponentUUIDs = new Set();
+        store.selectedComponentRectsMap = new Map();
+      });
+    },
+    /**
+     * Adds all of the components and their children ordered between the selected component and
+     * the active component (inclusive) to selectedComponentUUIDs.
+     */
+    addShiftSelectedComponentUUIDs: (selectedComponent: ComponentState) => {
+      const selectedUUID = selectedComponent.uuid;
+      const {
+        activeComponentUUID,
+        getActivePageState,
+        addSelectedComponentUUID,
+      } = get();
+      const activePageState = getActivePageState();
+      const componentTree = activePageState?.componentTree;
+      if (selectedUUID === activeComponentUUID) return;
+      if (!componentTree) {
+        throw new Error(
+          `Error adding components: component tree is undefined.`
+        );
+      }
+      if (!activeComponentUUID) {
+        throw new Error(
+          `Error adding components: active component is undefined.`
+        );
+      }
+      const targetComponentUUIDs = [selectedUUID, activeComponentUUID] as [
+        string,
+        string
+      ];
+      const lowestParentUUID = ComponentTreeHelpers.getLowestParentUUID(
+        ...targetComponentUUIDs,
+        componentTree
+      );
+
+      let selected = false;
+      ComponentTreeHelpers.mapComponentTreeParentsFirst(componentTree, (c) => {
+        if (c.parentUUID !== lowestParentUUID) return;
+        const descendants = ComponentTreeHelpers.getDescendants(
+          c,
+          componentTree
+        );
+        const targetIsParent = targetComponentUUIDs.includes(c.uuid);
+        const targetInDescendants = descendants.some((d) =>
+          targetComponentUUIDs.includes(d.uuid)
+        );
+        if (targetIsParent || targetInDescendants) {
```

Comment:
does this outside check do anything?

Reply:
The outside check `targetIsParent || targetInDescendants` checks whether we've reached the beginning or end of the range we need to select (between the active component and the clicked component). The inside check `(targetIsParent && targetInDescendants)` checks whether the start is the parent and the end is one of the children or vice versa because that would mean that we can just add everything in that branch and then stop.

Reply:
The way I just typed that was quite confusing lol, LMK if you need clarification.